[EME] Modify decryptor to always assume key ids are available.

charlie-ht opened the pull request:
Ideally, I would like the opencdm server to offer a blocking decrypt
method (with a configurable timeout) and then removing the keyReceived
logic, this is a minimal patch just to start looking up sessions with
key ids, but there's still a lot of code I don't think is necessary
assuming the above decrypt method which didn't end up being implemented
in the framework.

charlie-ht committed:
[EME] Modify decryptor to always assume key ids are available.

Ideally, I would like the opencdm server to offer a blocking decrypt
method (with a configurable timeout) and then removing the keyReceived
logic, this is a minimal patch just to start looking up sessions with
key ids, but there's still a lot of code I don't think is necessary
assuming the above decrypt method which didn't end up being implemented
in the framework.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
```diff
@@ -272,8 +272,18 @@ static GstFlowReturn webkitMediaCommonEncryptionDecryptTransformInPlace(GstBaseT
+    keyIDBuffer = gst_value_get_buffer(value);
+    if (gst_debug_category_get_threshold(GST_CAT_DEFAULT) >= GST_LEVEL_MEMDUMP) {
+        GstMapInfo map;
+        gst_buffer_map (keyIDBuffer, &map, GST_MAP_READ);
```
I guess we should check for the mapping to be correct before using that memory as it is done in other places.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -192,7 +193,13 @@ static bool webKitMediaOpenCDMDecryptorDecrypt(WebKitMediaCommonEncryptionDecryp
+    GstMapInfo keyIDBufferMap;
+    if (!gst_buffer_map(keyIDBuffer, &keyIDBufferMap, GST_MAP_READ)) {
+        gst_buffer_unmap(ivBuffer, &ivMap);
+        gst_buffer_unmap(buffer, &map);
```
I have discussed at some point with Phil that this kind of code is ugly and it would be awesome to create something like GstSmartMap that could handle this code gracefully. Because even if we are careful, we are (still) humans and we are going to mess up at some point.

This could be also good for upstream and bring it down.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -256,8 +264,9 @@ static bool webKitMediaOpenCDMDecryptorDecrypt(WebKitMediaCommonEncryptionDecryp
-        GST_TRACE_OBJECT(self, "decrypting");
-        if ((errorCode = priv->m_openCdm->Decrypt(map.data, static_cast<uint32_t>(map.size),
+        GST_TRACE_OBJECT(self, "decrypting (no subsamples)");
+        auto decryptor = std::make_unique<media::OpenCdm>(keyIDBufferMap.data, keyIDBufferMap.size);
```
Condidering that these are the two only places where priv->m_openCdm is used, I would just remove that attribute.

calvaris reviewed (changes requested):
I won't prevent landing this patch if you don't fix those things upstream but I would strongly encourage you to do so.

The only thing I would consider mandatory here would be removing the opencdm instance from the priv structure and the UNUSED.

Anyway, very good job!

charlie-ht committed:
[GStreamer] Add GstBufferMapped abstraction.

There are some similar abstractions I noticed while doing this called
`mapGstBuffer` and friends, which still seem to have a valid use-case:
wanting a buffer that is mapped for a longer lifetime without have to
keep track of the map infos separately. They could be subsumed by this
abstraction, but everytime they need to write to the buffer, they'd have
to remap the memory blocks.

This abstraction is more for one-short reads and writes saving the user
from remembering to unmap the buffer.

charlie-ht commented:
I've added that abstraction downstream for now, I don't want to deal with the conflicts between upstream at this time.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -75,4 +75,33 @@ inline GstClockTime toGstClockTime(const MediaTime &mediaTime)
+        : m_buffer(buffer)
+    {
+        if (gst_buffer_map(m_buffer, &m_info, flags))
+            m_isValid = true;
```
I think here you can do m_isValid = gst_buffer_map(m_buffer, &m_info, flags);

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -75,4 +75,33 @@ inline GstClockTime toGstClockTime(const MediaTime &mediaTime)
+    // Unfortunately, GST_MAP_READWRITE is defined out of line from the MapFlags
+    // enum, and C++ is careful to not implicity convert it to an enum.
+    explicit GstMappedBuffer(GstBuffer* buffer, int flags)
+        : GstMappedBuffer(buffer, static_cast<GstMapFlags>(flags)) {}
```
I would leave a space between { }

calvaris commented on Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
```diff
@@ -103,18 +103,16 @@ void InbandTextTrackPrivateGStreamer::notifyTrackOfSample()
-        gboolean ret = gst_buffer_map(buffer, &info, GST_MAP_READ);
-        ASSERT(ret);
-        if (!ret) {
+        GstMappedBuffer bufferMapped(buffer, GST_MAP_READ);
```
This is just a big IF coming from a non-native English speaker to a native English speaker. Shouldn't all these cases have the "mapped" in advance, like mappedBuffer, mappedInitData, mappedKeyIDBuffer, etc?

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -30,6 +30,7 @@
 #include <wtf/text/WTFString.h>
 #include <wtf/Lock.h>
 #include <wtf/PrintStream.h>
+#include <GStreamerCommon.h>
```
This has to go between CDMOpenCDM.h and open_cdm.h

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -256,27 +251,20 @@ static bool webKitMediaOpenCDMDecryptorDecrypt(WebKitMediaCommonEncryptionDecryp
-    gst_buffer_unmap(buffer, &map);
-    gst_buffer_unmap(ivBuffer, &ivMap);
-    return returnValue;
+    return true;
```
Now, we're talking...

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -179,39 +180,35 @@ static bool webKitMediaOpenCDMDecryptorAttemptToDecryptWithLocalInstance(WebKitM
-    GstMapInfo keyIDBufferMap;
-    if (!gst_buffer_map(keyIDBuffer, &keyIDBufferMap, GST_MAP_READ)) {
-        gst_buffer_unmap(ivBuffer, &ivMap);
-        gst_buffer_unmap(buffer, &map);
```
Frak yeah...

calvaris reviewed (changes requested):
Some style things, most of them minor, one I can't dismiss cause the style checker would complain

calvaris reviewed (approved):
I strongly encourage you to remove the unused opencdm attribute from the priv in the opencdm decryptor class before merging this.

charlie-ht commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitCommonEncryptionDecryptorGStreamer.cpp:
```diff
@@ -272,8 +272,20 @@ static GstFlowReturn webkitMediaCommonEncryptionDecryptTransformInPlace(GstBaseT
+    }
+
+    keyIDBuffer = gst_value_get_buffer(value);
+    if (gst_debug_category_get_threshold(GST_CAT_DEFAULT) >= GST_LEVEL_MEMDUMP) {
```
What's the incantation for that?

calvaris replied:
```
GST_DISABLE_GST_DEBUG

#define GST_DISABLE_GST_DEBUG 1
Configures the inclusion of the debugging subsystem
```

charlie-ht replied:
Thanks, the opencdm private member is still used for synchronising session availability with the server side, even though it shouldn't be needed, it is, and waiting on the framework opencdm to manage session availability

calvaris commented:
```
GST_DISABLE_GST_DEBUG

#define GST_DISABLE_GST_DEBUG 1
Configures the inclusion of the debugging subsystem
```
